rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+
+### Representing the runtime table
+
+The runtime table can be considered as a 'mask' that we apply to the fixed
+table to introduce proof-specific data at runtime.
+
+We make the simplifying assumption that an entry in the runtime table has
+exactly 1 'fixed' entry, which can be used for indexing (or set to `0` if
+unused), and that the table has a fixed width of `X`. We can pad any narrower
+entries in the table to the full width `X` in the same way as the fixed tables
+above.
+
+We represent the masked values in the runtime table as `RuntimeTable(i)` for
+`1 <= i < X`.
+
+To specify whether a runtime table entry is applicable in the `i`th row, we use
+a selector polynomial `RuntimeTableSelector`. In order to reduce the polynomial
+degree of later checks, we add a constraint to check that the runtime entry is
+the zero vector wherever the selector is zero:
+```
+RuntimeTableSelector * sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+ = sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
+```
+
+This gives the combined entry of the fixed and runtime tables as
+```
+sum[i=0, W, FixedTable(i) * joint_combiner^i]
++ sum[i=1, X, RuntimeTable(i) * joint_combiner^i]
++ TableID * joint_combiner^max_joint_size
+```
+where we assume that `RuntimeTableSelector * FixedTable(i) = 0` for
+`1 <= i < W` via our simplifying assumption above. We compute this as a
+polynomial `FixedAndRuntimeTable`, giving its evaluations and openings as part
+of the proof, and asserting that each entry matches the above.
+
+The `RuntimeTableSelector` polynomial will be fixed at constraint system time,
+and we can avoid opening it. However, we must provide openings for
+`RuntimeTable(i)`.
+
+The `joint_combiner` should be sampled using randomness that depends on the
+individual `RuntimeTable(i)`s, to ensure that the runtime values cannot be
+chosen with prior knowledge `joint_combiner`.
+
+### Representing the side-loaded tables
+
+The side-loaded tables can be considered as additional 'masks' that we apply to
+the combined fixed and runtime table, to introduce this side-loaded data.
+
+Again, we make the simplifying assumption that an entry from a side-loaded
+table has exactly 1 'fixed' entry, which can be used for indexing (or set to
+`0` if unused). We also assume that at most 1 side-loaded table contributes to
+each entry.
+
+In order to compute the lookup multiset inclusion argument, we also need the
+combined table of all 3 table kinds, which we will call `LookupTable`. We can
+calculate the contribution of the side-loaded table for each lookup as
+```
+LookupTable - FixedAndRuntimeTable
+```
+
+We will represent the polynomials for the side-loaded tables as
+`SideLoadedTable(i, j)`, where `i` identifies the table, and `j` identifies the
+column within that table. We also include polynomials `SideLoadedCombined(i)`
+in the proof, along with their openings, where
+```
+SideLoadedCombined(i)
+  = sum[j=1, ?, SideLoadedTable(i, j) * joint_combiner^(j-1)]
```

Comment:
why `j-1` and not `j`?

Reply:
re-reading this it still doesn't make sense to me, j should start at 0 here

Reply:
Correct, yes. The intention of the first column is as a mix-in so that you have (the possibility of) an indexed lookup a la memory accesses. You could set it to `0` if you don't care about the contents.

By doing it this way, we can load values into the side-loaded table without needing to care about their fixed argument -- since it's already implicit in the permutation argument.